Reporting work using Composition

akbargumbira opened the pull request:
* Fix Legend Issue in QGIS > 2.6
* Create  a new module map_report to handle reporting using composition.

Signed-off-by: Akbar Gumbira <akbargumbira@gmail.com>

akbargumbira commented:
Please don't merge this yet, I still want to:
- Remove the old safe.report.map
- Remove the usage of Map class in batch dialog to make a report and use the new reporting class
- Use the new reporting class in Impact Merge Dialog

timlinux commented on safe/gui/widgets/dock.py:
```diff
@@ -62,7 +62,7 @@
     FunctionOptionsDialog)
 from safe.common.utilities import temp_dir
-from safe.common.exceptions import ReadLayerError
+from safe.common.exceptions import ReadLayerError, LoadingTemplateError
```
TemplateLoadingError

timlinux commented on safe/common/exceptions.py:
```diff
@@ -288,6 +288,11 @@ class UnsupportedProviderError(Exception):
     pass
 
 
+class LoadingTemplateError(Exception):
```
TemplateLoadingError

timlinux commented on safe/gui/widgets/dock.py:
```diff
@@ -1478,8 +1479,20 @@ def restore_function_state(self, original_function):
+            button pressed."""
+        # Check if selected layer is valid
+        impact_layer = self.iface.activeLayer()
+        if impact_layer is None:
```
Maybe we should also check the keywords here to make sure it really is an impact summary layer.

akbargumbira replied:
right, will do

timlinux commented on safe/gui/widgets/dock.py:
```diff
@@ -1488,178 +1501,180 @@ def print_map(self):
+            return
+
+        # Print Impact Table
+        table_file_name = os.path.splitext(map_pdf_path)[0] + '_table.pdf'
```
TODO: Print the table into the same layout as the map....

akbargumbira replied:
ok will do now, can you point me to QGIS doc to this new feature?

timlinux replied:
http://docs.qgis.org/2.6/en/docs/user_manual/print_composer/print_composer.html#frames

Actually the feature is quite a old already - we just didnt take advantage of it yet....

timlinux replied:
I should add that early versions also had issues which is why we didnt use it - I think 2.6 improved things to prevent text being cut off on pagination etc.

timlinux commented on safe/report/map.py:
```diff
@@ -88,6 +90,15 @@ def set_impact_layer(self, layer):
 
+    def set_legend_layers(self, legend_layers):
+        """Set the layers that will be shown in the legend.
+        .. note: Added in 2.2 to cater the obsolescence of legend.model() in
```
.. note::

timlinux commented on safe/report/map.py:
```diff
@@ -88,6 +90,15 @@ def set_impact_layer(self, layer):
+    def set_legend_layers(self, legend_layers):
+        """Set the layers that will be shown in the legend.
+        .. note: Added in 2.2 to cater the obsolescence of legend.model() in
+        QGIS 2.6. The layers must be an instance of QgsLayerTreeGroup.
```
Needs to be indented more to be in the sphinx note box\

akbargumbira replied:
Don't bother with this. I am going to remove safe.report.map

timlinux commented on safe/report/map.py:
```diff
@@ -88,6 +90,15 @@ def set_impact_layer(self, layer):
+        self.legend_layers = legend_layers
+
     def set_north_arrow_image(self, north_arrow_path):
         """Set image that will be used as organisation logo in reports.
```
s/organisation logo / north arrow

akbargumbira replied:
Don't bother with this. I am going to remote safe.report.map. I just need to change its usages using new class

timlinux commented on safe/report/map_report.py:
```diff
@@ -0,0 +1,434 @@
+        return self._org_logo
+
+    @org_logo.setter
+    def org_logo(self, logo):
```
organisation_logo

timlinux commented on safe/report/map_report.py:
```diff
@@ -0,0 +1,434 @@
+            self._safe_logo = default_organisation_logo_path()
+
+    @property
+    def org_logo(self):
```
organisation_logo

timlinux commented on safe/report/map_report.py:
```diff
@@ -0,0 +1,434 @@
+        return self._extent
+
+    @extent.setter
+    def extent(self, extent):
```
What about some sanity checking like you have done for organisation_logo. We can check that the extent is valid, not None etc

timlinux commented:
Thanks very much for this work @akbargumbira +1 from me to merge when you are ready

akbargumbira commented:
Hi @ismailsunni and @borysiasty, I replaced the class Map by MapReport class (and removed the Map class) in this commit https://github.com/akbargumbira/inasafe/commit/4b28b54acb00416ac34693ec597f8019e7ccec0b. I updated the AnalysisHandler class to use this for printing the impact. I have tried it in Wizard Dialog and it's working. It's a duplicate of dock though, so they need to be refactored.

akbargumbira commented:
Hi @timlinux, I managed to use composition to print impact table. The output looks like this:
![composer_print](https://cloud.githubusercontent.com/assets/992519/5681465/5afaeafa-984b-11e4-878b-4a158f83ed79.png)

The table and the map pdf is still separated. The reason to separate that is that the template for the map might have landscape. The table does not look good on landscape and we can not print different paper orientation on different pages within one composition.

This PR will close:
- Legend layout not appearing in QGIS 2.6 https://github.com/AIFDR/inasafe/issues/1328
- PDF report table does not come with the style https://github.com/AIFDR/inasafe/issues/1115
- Refactor safe_qgis.report Modules https://github.com/AIFDR/inasafe/issues/923

If you are happy with this, I will add some tests, clean some codes, and merge it on Monday

akbargumbira commented:
Hi @timlinux, I am merging this. Some things that you think still are still not done can be issued in other tickets
